File: dotnet/src/Grounding/Ground.cs
using Grounding.Isaacus;

namespace Grounding;

/// <summary>What text the verdict classifier reads.</summary>
public enum ClassifierInput { Source, Candidates }

/// <summary>Retrieval backend: build once per request, reuse across claims.</summary>
public interface IRetriever
{
    Task<List<Candidate>> CandidatesAsync(string claim, int topK, CancellationToken ct);
    string SourceText(string sourceId);
}

/// <summary>
/// Chunk each source once; rerank the union of chunks per claim. Port of
/// SemchunkRetriever in src/grounding/ground.py. A claim is never grounded by the
/// chunk that restates it verbatim, nor by any later chunk of the same document.
/// </summary>
public sealed class SemchunkRetriever : IRetriever
{
    private readonly IsaacusClient _client;
    private readonly Dictionary<string, string> _text = new();
    private readonly string? _selfId;
    // (source_id, chunk_index, text) for every chunk, in document order.
    private readonly List<(string Sid, int Idx, string Text)> _chunks = new();

    public SemchunkRetriever(IsaacusClient client, IEnumerable<Source> sources,
        string? selfExcludeSourceId = null)
    {
        _client = client;
        _selfId = selfExcludeSourceId;
        foreach (var s in sources)
        {
            _text[s.Id] = s.Text;
            int i = 0;
            foreach (var ch in Chunk.ChunkSource(s.Text))
                _chunks.Add((s.Id, i++, ch));
        }
    }

    public string SourceText(string sourceId) => _text.GetValueOrDefault(sourceId, "");

    private List<(string Sid, int Idx, string Text)> Eligible(string claim)
    {
        if (_selfId is null)
            return _chunks;
        string norm = TextUtil.Norm(claim);
        int? cutoff = null;
        foreach (var (sid, idx, ch) in _chunks)
        {
            if (sid == _selfId && norm.Length > 0 && TextUtil.Norm(ch).Contains(norm))
                cutoff = cutoff is null ? idx : Math.Min(cutoff.Value, idx);
        }
        if (cutoff is null)
            return _chunks;
        return _chunks.Where(c => !(c.Sid == _selfId && c.Idx >= cutoff.Value)).ToList();
    }

    public async Task<List<Candidate>> CandidatesAsync(string claim, int topK, CancellationToken ct)
    {
        var pool = Eligible(claim);
        if (pool.Count == 0)
            return new List<Candidate>();
        var texts = pool.Select(c => c.Text).ToList();
        var resp = await _client.RerankAsync(claim, texts, ct);
        var outp = new List<Candidate>();
        foreach (var r in resp.Results.Take(topK))
        {
            var (sid, idx, ch) = pool[r.Index];
            outp.Add(new Candidate(ch, sid, idx, TextUtil.Round4(r.Score)));
        }
        return outp;
    }
}

/// <summary>Result of grounding one claim (mirrors the Grounding dataclass).</summary>
public sealed record Grounding(
    List<Candidate> Candidates,
    Candidate? Seed,
    double Relevance,
    SupportingSpan? Span,
    double Inextract,
    double AnswerScore,
    double PSupport,
    double PContra,
    SupportingSpan? ClsSpan,
    string Verdict,
    double Confidence);

/// <summary>
/// Unified per-claim grounding core: retrieve → extract → classify → verdict.
/// Port of ground.py (semchunk path; ILDGS backend deferred).
/// </summary>
public sealed class Grounder
{
    public const int RerankK = 3;

    private readonly IsaacusClient _client;
    private readonly Classify _classify;
    private readonly Extract _extract;

    public Grounder(IsaacusClient client, Classify classify, Extract extract)
    {
        _client = client;
        _classify = classify;
        _extract = extract;
    }

    /// <summary>Pick the seed sentence most relevant to the claim, via the reranker.</summary>
    private async Task<SupportingSpan?> BestLineAsync(string claim, string passage, CancellationToken ct)
    {
        var sents = TextUtil.SplitSentences(passage)
            .Select(s => s.Text.Trim())
            .Where(t => t.Length > 15)
            .ToList();
        if (sents.Count == 0)
            return null;
        var resp = await _client.RerankAsync(claim, sents, ct);
        if (resp.Results.Count == 0)
            return null;
        var top = resp.Results[0];
        string sent = sents[top.Index];
        int start = passage.IndexOf(sent, StringComparison.Ordinal);
        if (start < 0) start = 0;
        return new SupportingSpan(sent, start, start + sent.Length, TextUtil.Round4(top.Score));
    }

    public async Task<Grounding> GroundAsync(
        string claim, IRetriever retriever, int topK = RerankK,
        ClassifierInput classifierInput = ClassifierInput.Source,
        bool withLine = false, CancellationToken ct = default)
    {
        var candidates = await retriever.CandidatesAsync(claim, topK, ct);
        if (candidates.Count == 0)
            return new Grounding(new(), null, 0.0, null, 1.0, 0.0, 0.0, 0.0, null, "unaddressed", 0.0);

        var seed = candidates[0];
        double relevance = seed.Score;
        string srcText = retriever.SourceText(seed.SourceId);
        if (srcText.Length == 0) srcText = seed.Text;

        // inextract feeds the verdict (Rule 3 guard); always from the extractor.
        var (span, inextract) = await _extract.ExtractSpanAsync(
            claim, new[] { seed.Text }, relevanceConfirmed: true, ct: ct);
        double answerScore = span?.Score ?? 0.0;

        // Displayed quote: extractor span is unreliable on declarative claims, so
        // for the UI use the reranked best sentence of the seed instead.
        if (withLine)
        {
            var line = await BestLineAsync(claim, seed.Text, ct);
            if (line is not null)
            {
                span = line;
                answerScore = line.Score;
            }
        }

        var candTexts = candidates.Select(c => c.Text).ToList();
        var clsTexts = classifierInput == ClassifierInput.Source
            ? new List<string> { srcText }
            : candTexts;
        var scores = await _classify.ClassifyScoresAsync(claim, clsTexts, ct);

        var (verdict, confidence) = Classify.VerdictFromScores(
            scores.PSupport, scores.PContra, inextract);

        return new Grounding(candidates, seed, relevance, span, inextract, answerScore,
            scores.PSupport, scores.PContra, scores.Span, verdict, confidence);
    }

    /// <summary>Build the API ClaimResult from a Grounding (mirrors to_claim_result).</summary>
    public static ClaimResult ToClaimResult(Claim claim, Grounding g, bool includePassages = false)
    {
        if (g.Seed is null)
            return new ClaimResult(claim.Text, "unaddressed", 0.0, RangeRef: claim.RangeRef);

        SupportingSpan? finalSpan;
        if (g.Verdict == "unaddressed")
            finalSpan = null;
        else if (g.Verdict == "weak")
            finalSpan = g.AnswerScore > 0.3 ? g.Span : g.ClsSpan;
        else
            finalSpan = g.Span ?? g.ClsSpan;

        var evidence = new List<EvidenceItem>
        {
            new(g.Seed.SourceId, g.Seed.Text, Relation: "seed", Score: g.Seed.Score),
        };

        var passages = new List<RankedPassage>();
        if (includePassages)
        {
            for (int i = 0; i < g.Candidates.Count; i++)
            {
                var c = g.Candidates[i];
                bool isSeed = i == 0;
                passages.Add(new RankedPassage(
                    c.SourceId, c.Score, c.Text,
                    Line: isSeed && g.Span is not null ? g.Span.Text : null,
                    LineStart: isSeed && g.Span is not null ? g.Span.Start : null,
                    LineEnd: isSeed && g.Span is not null ? g.Span.End : null));
            }
        }

        return new ClaimResult(
            claim.Text, g.Verdict, g.Confidence, RangeRef: claim.RangeRef,
            Span: finalSpan, SourceId: g.Seed.SourceId, Evidence: evidence, Passages: passages);
    }
}

File: python/src/grounding/ground.py
"""Unified per-claim grounding core.

One path shared by production (pipeline.py) and the eval harnesses
(calibrate.py, run_eval.py) so they cannot diverge again. Retrieval backend is
pluggable; the verdict logic and classifier call are identical across backends.

Retrieval backends (build once per request, reused across claims):
  SemchunkRetriever (default) — chunk each source with semchunk, rerank the
    union of chunks across sources, return the top_k as Candidates.
  IldgsRetriever (opt-in, GROUNDING_BACKEND=ildgs) — the enrich → graph → embed
    → route path, wrapped to emit the same Candidate shape.

classifier_input selects what the verdict classifier reads:
  "source"     — full source text of the seed (auto-chunked; robust to silence,
                 the current production behaviour).
  "candidates" — only the reranked top_k passages (Workstream-1 experiment).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal, Protocol

from .chunk import chunk_source
from .classify import classify_scores, verdict_from_scores
from .client import get_client
from .extract import extract_span
from .models import Candidate, ClaimResult, EvidenceItem, RankedPassage, SupportingSpan
from .segment import _nlp

ClassifierInput = Literal["source", "candidates"]


def _norm(text: str) -> str:
    return " ".join(text.split()).lower()


def _best_line(claim: str, passage: str) -> SupportingSpan | None:
    """Pick the single sentence of `passage` most relevant to `claim`, via the
    reranker, as the human-facing supporting quote.

    The answer-extractor is extractive QA tuned for *questions*; on declarative
    NLI-style claims it returns near-zero-score fragments (often a heading), so
    it is unreliable for the displayed line. Sentence-level reranking gives the
    exact operative sentence (for a contradiction, the clause it conflicts with),
    with char offsets back into `passage`.
    """
    sents = [s.text.strip() for s in _nlp()(passage).sents if len(s.text.strip()) > 15]
    if not sents:
        return None
    resp = get_client().rerankings.create(
        model="kanon-2-reranker", query=claim, texts=sents,
    )
    if not resp.results:
        return None
    top = resp.results[0]
    sent = sents[top.index]
    start = passage.find(sent)
    if start < 0:
        start = 0
    return SupportingSpan(text=sent, start=start, end=start + len(sent),
                          score=round(top.score, 4))


# ── Retrievers ────────────────────────────────────────────────────────────────

class Retriever(Protocol):
    def candidates(self, claim: str, *, top_k: int = 3) -> list[Candidate]: ...
    def source_text(self, source_id: str) -> str: ...


class SemchunkRetriever:
    """Chunk each source once; rerank the union of chunks per claim.

    `self_exclude_source_id` marks the main document: a claim is never grounded
    by the chunk that restates it verbatim, nor by any later chunk of that same
    document (mirrors pipeline._self_exclude — a claim may only be grounded by
    content preceding it, or by other sources).
    """

    def __init__(self, sources, *, self_exclude_source_id: str | None = None):
        self._text = {s.id: s.text for s in sources}
        self._self_id = self_exclude_source_id
        # (source_id, chunk_index, text) for every chunk, in document order.
        self._chunks: list[tuple[str, int, str]] = []
        for s in sources:
            for i, ch in enumerate(chunk_source(s.text)):
                self._chunks.append((s.id, i, ch))

    def source_text(self, source_id: str) -> str:
        return self._text.get(source_id, "")

    def _eligible(self, claim: str) -> list[tuple[str, int, str]]:
        if not self._self_id:
            return self._chunks
        norm = _norm(claim)
        cutoff: int | None = None
        for sid, idx, ch in self._chunks:
            if sid == self._self_id and norm and norm in _norm(ch):
                cutoff = idx if cutoff is None else min(cutoff, idx)
        if cutoff is None:
            return self._chunks
        return [c for c in self._chunks
                if not (c[0] == self._self_id and c[1] >= cutoff)]

    def candidates(self, claim: str, *, top_k: int = 3) -> list[Candidate]:
        pool = self._eligible(claim)
        if not pool:
            return []
        texts = [c[2] for c in pool]
        resp = get_client().rerankings.create(
            model="kanon-2-reranker", query=claim, texts=texts,
        )
        out: list[Candidate] = []
        for r in resp.results[:top_k]:
            sid, idx, ch = pool[r.index]
            out.append(Candidate(text=ch, source_id=sid, chunk_index=idx,
                                 score=round(r.score, 4)))
        return out


class IldgsRetriever:
    """Adapter over the ILDGS graph path so it emits Candidates like semchunk.

    Built from a prepared SessionGraph + segment embeddings (shared across claims
    by the caller). Cross-reference/term routing is preserved for evidence.
    """

    def __init__(self, graph, seg_embeddings, *, exclude_fn=None):
        self._graph = graph
        self._emb = seg_embeddings
        self._exclude_fn = exclude_fn  # claim -> frozenset[node_key]

    def source_text(self, source_id: str) -> str:
        return self._graph.source_text.get(source_id, "")

    def candidates(self, claim: str, *, top_k: int = 3) -> list[Candidate]:
        from .retrieve import retrieve  # local import: optional backend
        exclude = self._exclude_fn(claim) if self._exclude_fn else frozenset()
        ranked = retrieve(self._graph, claim, self._emb, exclude=exclude,
                          rerank_k=top_k)
        return [Candidate(text=r.text, source_id=r.source_id, chunk_index=0,
                          score=r.score) for r in ranked]


# ── Per-claim grounding ───────────────────────────────────────────────────────

@dataclass
class Grounding:
    candidates: list[Candidate]
    seed: Candidate | None
    relevance: float
    span: SupportingSpan | None
    inextract: float
    answer_score: float
    p_support: float
    p_contra: float
    cls_span: SupportingSpan | None
    verdict: str
    confidence: float
    # Workstream-1 experiment fields (None unless also_reranked=True):
    p_support_rr: float | None = None
    p_contra_rr: float | None = None


def ground(
    claim: str,
    retriever: Retriever,
    *,
    top_k: int = 3,
    classifier_input: ClassifierInput = "source",
    also_reranked: bool = False,
    with_line: bool = False,
) -> Grounding:
    """Retrieve → extract → classify → verdict for one claim. No graph building.

    `with_line=True` replaces the displayed supporting span with the best seed
    sentence (reranked) — a correct human-facing quote. Off by default so the
    eval harness keeps the extractor's answer_score signal unchanged.
    """
    candidates = retriever.candidates(claim, top_k=top_k)
    if not candidates:
        return Grounding([], None, 0.0, None, 1.0, 0.0, 0.0, 0.0, None,
                         "unaddressed", 0.0)

    seed = candidates[0]
    relevance = seed.score
    src_text = retriever.source_text(seed.source_id) or seed.text

    # inextract feeds the verdict (Rule 3 guard); always from the extractor.
    span, inextract = extract_span(claim, [seed.text], relevance_confirmed=True)
    answer_score = span.score if span else 0.0

    # The displayed quote: the extractor span is unreliable on declarative claims,
    # so for the UI use the reranked best sentence of the seed instead.
    if with_line:
        line = _best_line(claim, seed.text)
        if line is not None:
            span = line
            answer_score = line.score

    cand_texts = [c.text for c in candidates]
    cls_texts = [src_text] if classifier_input == "source" else cand_texts
    p_support, p_contra, cls_span = classify_scores(claim, cls_texts)

    p_support_rr = p_contra_rr = None
    if also_reranked:
        # The other signal set, for the offline classifier-input comparison.
        if classifier_input == "source":
            p_support_rr, p_contra_rr, _ = classify_scores(claim, cand_texts)
        else:
            ps2, pc2, _ = classify_scores(claim, [src_text])
            p_support_rr, p_contra_rr = p_support, p_contra
            p_support, p_contra = ps2, pc2  # keep "source" as the primary fields

    verdict, confidence = verdict_from_scores(p_support, p_contra, inextract)
    return Grounding(candidates, seed, relevance, span, inextract, answer_score,
                     p_support, p_contra, cls_span, verdict, confidence,
                     p_support_rr, p_contra_rr)


def to_claim_result(claim, g: Grounding, *, include_passages: bool = False) -> ClaimResult:
    """Build the API ClaimResult from a Grounding (mirrors pipeline rendering)."""
    if g.seed is None:
        return ClaimResult(text=claim.text, range_ref=claim.range_ref,
                           verdict="unaddressed", confidence=0.0)

    if g.verdict == "unaddressed":
        final_span = None
    elif g.verdict == "weak":
        final_span = g.span if g.answer_score > 0.3 else g.cls_span
    else:
        final_span = g.span or g.cls_span

    evidence = [EvidenceItem(source_id=g.seed.source_id, relation="seed",
                             text=g.seed.text, score=g.seed.score)]
    passages: list[RankedPassage] = []
    if include_passages:
        for i, c in enumerate(g.candidates):
            is_seed = i == 0
            passages.append(RankedPassage(
                source_id=c.source_id, score=c.score, passage=c.text,
                line=g.span.text if (is_seed and g.span) else None,
                line_start=g.span.start if (is_seed and g.span) else None,
                line_end=g.span.end if (is_seed and g.span) else None,
            ))

    return ClaimResult(
        text=claim.text, range_ref=claim.range_ref, verdict=g.verdict,
        confidence=g.confidence, span=final_span, source_id=g.seed.source_id,
        evidence=evidence, passages=passages,
    )
